openedx/openedx-platform
# [ENT-17] Implement basics of UserDataSharingConsentAudit

This pull request brings into place the very basics of ENT-17, in the form of a consent audit model, as well as a historical record. Due to the necessity of manual testing, it also brings the basic form of additional views and form changes that are necessary to take advantage of the model change.

**JIRA tickets**: Implements ENT-17

**Discussions**: Discussion and discovery on ENT-24

**Dependencies**: None

**Sandbox URL**: http://pr13687.sandbox.opencraft.hosting/ (use standard `staff` account for admin access)

**Testing instructions**:
1. In the sandbox admin, verify that the TestShib SAML IdP is set to neither require nor request data sharing consent.
2. Register a new user, using the TestShib SAML IdP. Verify that the process is identical to the one that exists now.
3. Disconnect that user from the IdP and sign out.
4. Modify the TestShib SAML IdP setting to request data sharing consent.
5. Register a new user, using the TestShib SAML IdP, noting that a checkbox for consent is present, checking it.
6. In the admin panel, note that DataSharingConsentAudit object for this user notes that data sharing is permitted.
7. Disconnect this user from the IdP and sign out.
8. Register a new user, using the TestShib SAML IdP, noting that a checkbox for consent is present, but leaving it blank.
9. Sign out of this user, but do not disconnect it from the IdP.
10. In the admin panel, note that the DataSharingConsentAudit for this user notes that data sharing is not permitted.
11. In the admin panel, set the TestShib SAML IdP to require data sharing consent.
12. Attempt to sign in, using the TestShib SAML user account you registered previously. Note that it redirects you to a page stating that additional permissions are required, and that if you manually browse elsewhere, you are not logged in.
13. In the admin panel, remove the requirement for data sharing consent on the TestShib SAML IdP.
14. Log in with the TestShib user, noting that you can do so successfully.
15. Disconnect the user from the IdP and sign out.
16. In the admin panel, re-enable the requirement for data sharing consent on the TestShib SAML IdP.
17. Attempt to register a new user using the TestShib SAML IdP. Note that the checkbox for data sharing consent is present; note that if you attempt to register a user without checking it, you encounter an error message.
18. Note that once registered, the OpenEdX platform behaves normally.

**Author notes and concerns**:
1. There are a couple things included in this PR that are not required by ENT-17, but which were necessary to fully test behavior - for example, the addition of the data sharing consent checkbox to the registration form.

**Reviewers**
- [ ] @edx/devops 

**Settings**

``` yaml
EDXAPP_FEATURES:
  ENABLE_COMBINED_LOGIN_REGISTRATION: true
  ENABLE_THIRD_PARTY_AUTH: true
```
